qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -1156,26 +1212,73 @@ def block_for_results(self, timeout: Optional[float] = None) -> "DbExperimentDat
         Returns:
             The experiment data with finished jobs and post-processing.
         """
-        if self._analysis_futures:
-            self._wait_for_analysis(timeout)
-        else:
-            self._wait_for_jobs(timeout)
-        self._removed_done_futures()
+        start_time = time.time()
+        with self._job_futures.lock and self._analysis_futures.lock:
+            # Lock threads to get all current job and analysis futures
+            # at the time of function call and then release the lock
+            job_ids = self._job_futures.keys()
+            job_futs = self._job_futures.values()
+            analysis_ids = self._analysis_futures.keys()
+            analysis_futs = self._analysis_futures.values()
+
+        # Wait for futures
+        self._wait_for_futures(job_futs + analysis_futs, name="jobs and analysis", timeout=timeout)
+
+        # Clean up done job futures
+        num_jobs = len(job_ids)
+        for jid, fut in zip(job_ids, job_futs):
+            if (fut.done() and not fut.exception()) or fut.cancelled():
```

Comment:
The way futures exceptions are currently handled we need to keep the future object around to get the exception later, so these can't be deleted here.

There is no necessary requirement to delete futures at all, it's just for some internal cleanup to keep the futures lits from growing if you keep re-running analysis, since all these futures will be waited on in various locations.